components/support/rc_crypto/src/constant_time.rs

```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+
+/// Returns `Ok(())` if `a == b` and `Error` otherwise.
+/// The comparison of `a` and `b` is done in constant time with respect to the
+/// contents of each, but NOT in constant time with respect to the lengths of
+/// `a` and `b`.
+
+pub fn verify_slices_are_equal(a: &[u8], b: &[u8]) -> Result<()> {
+    if a.len() != b.len() {
+        return Err(ErrorKind::InternalError.into());
+    }
+    let mut result: u8 = 0;
+    for n in 0..a.len() {
+        result |= a[n] ^ b[n];
+    }
+    match result {
+        0 => Ok(()),
+        _ => Err(ErrorKind::InternalError.into()),
+    }
+}
```

Comment:
There's a lot involved in doing constant-time-eq right, including trying to prevent the optimizer from optimizing things away.  We may be better off taking a dependency on [constant_time_eq](https://crates.io/crates/constant_time_eq) for this, especially since that appears to be vendored in Firefox already.

Reply:
(Or see if we can import `NSS_SecureMemcmp` I suppose, but I'm assuming that you deliberately weren't doing that).

Reply:
Oh, either of these are better calls than my suggestion. I'm having github issues though.

Reply:
Having peeked at the source of `constant_time_eq` briefly, it probably has at least some of the problems with optimizing compilers that the linked `ring` issue mentions. Using functionality exposed by NSS would be best here, if it's possible.